### PR TITLE
fix(quic-transport): Send `RESET_STREAM` frame for transport handles that are dropped

### PR DESCRIPTION
**Pull request by DSharifi**
This PR adds a drop guard to wrap `quinn::SendStream` in our transport implementation when making RPC calls. The wrapper will send a `SendStream::reset` frame if it is dropped, which can happen if a transport client, such as P2P, cancels the RPC. Quinn sends a `finish` frame by default when dropping a stream, which can lead to the peer receiving the stream thinking a complete message was sent, and attempts to decode an incomplete message. This guard is used to send a reset frame instead, to signal to the peer that the transmission of the message was cancelled.

### CODE REVIEW

File: rs/p2p/quic_transport/src/connection_handle.rs [hunk 1]
```diff
@@ -194,3 +214,77 @@ async fn write_request(
...
+            .await
+            .unwrap();
+
+        drop(drop_guard);
+
+        let server_result = handler.await.unwrap();
+
+        assert_matches!(
+            server_result,
+            Err(ReadToEndError::Read(ReadError::Reset { .. }))
```
Comment (tthebst) [inline rs/p2p/quic_transport/src/connection_handle.rs:287]: Could you also test that return Ok. If finish and stop was called on the send stream?

Comment (DSharifi) [reply rs/p2p/quic_transport/src/connection_handle.rs:287]: What `stop` method are you referring to? Right now the receiver will also return an error if the sender sends a `finish` frame, and then drops the drop guard which sends a `reset` frame.

File: rs/p2p/quic_transport/BUILD.bazel [hunk 2]
```diff
@@ -62,6 +64,14 @@ rust_library(
     deps = DEPENDENCIES,
 )
 
+rust_test(
+    name = "quic_transport_unit",
```
Comment (tthebst) [inline rs/p2p/quic_transport/BUILD.bazel:68]:
```suggestion
    name = "quic_transport_test",
```

File: rs/p2p/quic_transport/src/connection_handle.rs [hunk 3]
```diff
@@ -21,6 +21,26 @@ use crate::{
...
+impl SendStreamDropGuard {
+    fn new(send_stream: SendStream) -> Self {
+        Self { send_stream }
+    }
+}
+
+impl Drop for SendStreamDropGuard {
+    fn drop(&mut self) {
+        // fails silently if the stream is already closed.
+        let _ = self.send_stream.reset(VarInt::from_u32(0));
```
Comment (tthebst) [inline rs/p2p/quic_transport/src/connection_handle.rs:40]: Note that error code zero means no error. See https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-03#section-12.3

> QUIC_STREAM_CANCELLED (0x80000006):  The stream was cancelled

Is probably more appropriate.

**Review by tthebst — APPROVED** [review]

File: rs/p2p/quic_transport/BUILD.bazel (hunk 2 above)
Comment (DSharifi) [reply rs/p2p/quic_transport/BUILD.bazel:68]: Seems like we use the "unit" and "integration" suffix for most of our tests. Wouldn't this break with that convention?

Comment (tthebst) [reply rs/p2p/quic_transport/BUILD.bazel:68]: I thought our convention is "test" and "integration" (which transforms to "integration_test")

**Pull request merged**